pkg/controllers/controllers.go

```diff
@@ -200,6 +235,28 @@ func RegisterControllers(ctx context.Context, m manager.Manager, controllers ...
 	return m
 }
 
+// StartCleanupWatcher monitors the signal channel for termination, closes the cleanupAsync channel when
+// there is a signal for pod termination, waits for the cleanup operation to complete, and then cancels all contexts
+// Only the leader will perform the cleanup operation
+func StartCleanupWatcher(ctx context.Context, cancel context.CancelFunc, elected <-chan struct{},
```

Comment:
I'm a little confused, even if the `controller-runtime` catches `SIGTERM` and closes the `Done` channel, that doesn't affect the parent context. 

Also, from what I can tell, `controller-runtime` does not do this by default. They have support using the `SetupSignalHandler` to be passed into the `mgr.Start` call but they don't catch these signals internally. Point to me a link if my understanding is wrong because I could definitely be missing something.

Reply:
As far as I can tell, the signal handling context is created [here](https://github.com/aws/karpenter/blob/b1ef730df63cea61a970fe3f4c046edbeec2925d/pkg/utils/injection/injection.go#L36), which is passed to manager.Start, which [returns](https://github.com/kubernetes-sigs/controller-runtime/blob/7399a3a595bf254add9d0c96c49af462e1aac193/pkg/manager/internal.go#L501) when the context is closed by the signal which is is the only thing that keeping main from exiting as it's the last thing called by `controllers.Initialize` [here](https://github.com/aws/karpenter/blob/b1ef730df63cea61a970fe3f4c046edbeec2925d/pkg/controllers/controllers.go#L160).

When I tested I didn't see the log line indicating that the cleanup functions had finished running.

Reply:
Ah, I completely missed that we used `signal.NewContext()` in `EnableInjectionOrDie`. Good catch!

Reply:
Discussed offline: cleanup is now going to be handled when there are no Provisioners managed by the controller. We will modify the infrastructure controller to run while there are > 0 Provisioners on the cluster and delete the queue when there are <= 0 Provisioners on the cluster